File: 백준/Gold/1647. 도시 분할 계획/도시 분할 계획.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

class Main{
    private static int n;
    private static int[] parents;

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer tokenizer = new StringTokenizer(br.readLine());

        n = Integer.parseInt(tokenizer.nextToken());
        int m = Integer.parseInt(tokenizer.nextToken());
        
        parents = new int[n + 1];
        for(int i=1 ; i<=n ; i++){
            parents[i] = i;
        }

        PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> o1[2] - o2[2]);
        while(m-- > 0){
            tokenizer = new StringTokenizer(br.readLine());
            int sour, des, weight;

            sour = Integer.parseInt(tokenizer.nextToken());
            des = Integer.parseInt(tokenizer.nextToken());
            weight = Integer.parseInt(tokenizer.nextToken());

            pq.add(new int[]{sour, des, weight});
        }

        System.out.println(mst(pq));
    }

    private static int mst(PriorityQueue<int[]> pq){   
        int ret = 0;

        int temp = 0;
        while(!pq.isEmpty()){
            int[] cur = pq.poll();
            int x = cur[0], y = cur[1], weight = cur[2];

            if(findParent(x) == findParent(y))
                continue;
            
            union(x, y);
            temp = weight;
            ret += weight;
        }

        return ret - temp;
    }
    
    private static int findParent(int x){
        if(x == parents[x])
            return x;
        return parents[x] = findParent(parents[x]);
    }

    private static void union(int x, int y){
        int px = findParent(x), py = findParent(y);

        if(px == py)
            return;

        parents[py] = px;
    }

}

File: 백준/Gold/1647. 도시 분할 계획/도시 분할 계획.py
from sys import stdin
from heapq import heappush, heappop

input = stdin.readline

n, m = map(int, input().split())

graph = []
parents = [i for i in range(n + 1)]

for sour, des, cost in [list(map(int, input().split())) for _ in range(m)]:
    heappush(graph, (cost, sour, des))

    
def find_parent(x):
    global parents

    if parents[x] == x:
        return x

    parents[x] = find_parent(parents[x])

    return parents[x]


def union(x, y):
    global parents

    if x == y:
        return

    parents[y] = x

answer = 0
temp = 0
while graph:
    cost, sour, des = heappop(graph)
    sour_parent, des_parent = find_parent(sour), find_parent(des)
    if sour_parent == des_parent:
        continue
    answer += cost
    temp = cost
    union(sour_parent, des_parent)

print(answer - temp)
